clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -211,6 +213,15 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
                     maxTimestamp = info.maxTimestamp;
                     shallowOffsetOfMaxTimestamp = info.shallowOffsetOfMaxTimestamp;
                 }
+            } else if (batchRetention == BatchRetention.RETAIN_EMPTY) {
+                if (batchMagic < RecordBatch.MAGIC_VALUE_V2)
+                    throw new IllegalStateException("Empty batches are only supported for magic v2 and above");
+
+                bufferOutputStream.maybeExpandBuffer(DefaultRecordBatch.RECORD_BATCH_OVERHEAD);
+                DefaultRecordBatch.writeEmptyHeader(bufferOutputStream.buffer(), batchMagic, batch.producerId(),
```

Comment:
> it might have been more consistent to preserve the base timestamp and max timestamp from the original batch as well

What's the disadvantage of just doing that?